Addition of GDExtension docs (#6212)
godotengine/godot-docs: tutorials/scripting/gdextension/gdextension_cpp_example.rst

Signals
-------

Last but not least, signals fully work in GDExtension as well. Having your extension
react to a signal given out by another object requires you to call ``connect``
on that object. We can't think of a good example for our wobbling Godot icon, we
would need to showcase a far more complete example.

This is the required syntax:

.. code-block:: C++

    some_other_node->connect("the_signal", this, "my_method");

Note that you can only call ``my_method`` if you've previously registered it in
your ``_bind_methods`` method.

Having your object sending out signals is more common. For our wobbling
Godot icon, we'll do something silly just to show how it works. We're going to
emit a signal every time a second has passed and pass the new location along.

In our ``gdexample.h`` header file, we need to define a new member ``time_emit``:

.. code-block:: C++

    ...
        float time_passed;
        float time_emit;
        float amplitude;
    ...

This time, the changes in ``gdexample.cpp`` are more elaborate. First,
you'll need to set ``time_emit = 0.0;`` in either our ``_init`` method or in our
constructor. We'll look at the other 2 needed changes one by one.

> In our ``_register_methods`` method, we need to declare our signal. This is done

Reviewer: ```suggestion
In our ``_bind_methods`` method, we need to declare our signal. This is done
```

I think you meant `_bind_methods`.

Author: Corrected

as follows:

.. code-block:: C++

    void GDExample::_bind_methods() {
        ClassDB::add_property("GDExample", PropertyInfo(Variant::FLOAT, "amplitude"), "set_amplitude", "get_amplitude");
        ClassDB::add_property("GDExample", PropertyInfo(Variant::FLOAT, "speed"), "set_speed", "get_speed");

        ADD_SIGNAL(MethodInfo("position_changed", PropertyInfo(Variant::OBJECT, "node"), PropertyInfo(Variant::VECTOR2, "new_pos")));
    }

> Here, our ``register_signal`` method can be a single call first taking the

Reviewer: ```suggestion
Here, our ``ADD_SIGNAL`` method can be a single call first taking the
```

ADD_SIGNAL?

Author: Corrected

libraries that can be compiled into your project stored in ``godot-cpp/bin/``.

.. note::

    You may need to add ``bits=64`` to the command on Windows or Linux.

Creating a simple plugin
------------------------

Now it's time to build an actual plugin. We'll start by creating an empty Godot
project in which we'll place a few files.

Open Godot and create a new project. For this example, we will place it in a
folder called ``demo`` inside our GDExtension module's folder structure.

In our demo project, we'll create a scene containing a Node called "Main" and
we'll save it as ``main.tscn``. We'll come back to that later.

Back in the top-level GDExtension module folder, we're also going to create a
subfolder called ``src`` in which we'll place our source files.

You should now have ``demo``, ``godot-cpp``, and ``src``
directories in your GDExtension module.

In the ``src`` folder, we'll start with creating our header file for the
GDExtension node we'll be creating. We will name it ``gdexample.h``:

.. code-block:: C++

    #ifndef GDEXAMPLE_H
    #define GDEXAMPLE_H

    #include <sprite2d.hpp>

    namespace godot {

    class GDExample : public Sprite2D {

>         GD_CLASS(GDExample, Sprite2D)

Reviewer: Hey, I think you meant `GDCLASS` here

>         GD_CLASS(GDExample, Sprite2D)

Reviewer: ```suggestion
        GDCLASS(GDExample, Sprite2D)
```

I think you meant `GDCLASS`. I got stuck here for a little while 🤓

Author: @yudi-azvd Thanks for noticing! It's updated now.

.. _doc_gdextension_cpp_example:

GDExtension C++ example

> ====================

Reviewer 2: Underline length isn't correct

Introduction
------------

The C++ bindings for GDExtension are built on top of the C GDExtension API

> and provide a nicer way to "extend" nodes in Godot using C++. This new system allows the 

Reviewer 3: "extend" any Class

Reviewer 3: I do not know if "any" is accurate but it's more then only the nodes.

Author: There is no `any`? or do you mean you want it more specific?

Reviewer 3: I see all the demo cases extending nodes. Maybe in the tutorial, there should be some reference that other things are also able to extend.

Author: Updated it with this: 
> nicer way to "extend" nodes and other built-in classes in Godot using C++.

extension of Godot to nearly the same level as statically linked C++ modules.

You can download the included example in the test folder of the godot-cpp repository `on
GitHub <https://github.com/godotengine/godot-cpp>`__.

Setting up the project
----------------------

There are a few prerequisites you'll need:

> - a Godot 4.x executable,

Reviewer 2: ```suggestion
- a Godot 4 executable,
```

- a C++ compiler,
- SCons as a build tool,
- a copy of the `godot-cpp
  repository <https://github.com/godotengine/godot-cpp>`__.

See also :ref:`Compiling <toc-devel-compiling>` as the build tools are identical
to the ones you need to compile Godot from source.

You can download these repositories from GitHub or let Git do the work for you.
Note that these repositories have different branches for different versions

> of Godot. GDExtension modules written will not work in older versions of Godot (3.x) 

Reviewer 2: ```suggestion
of Godot. GDExtension modules will not work in older versions of Godot (only Godot 4 and up) 
```

Author: I also removed the `modules` word as it just confuses the user (modules are a different thing entirely and shouldn't be used as synonyms for extensions)

> and vice versa, so make sure you download the correct branch.

Reviewer 2: This paragraph refers to "repositories", plural, whereas the instructions immediately above only talk about a single repository.

.. note::

    `GDExtension <https://godotengine.org/article/introducing-gd-extensions>`__
    has been merged in the ``master`` branch of godot-cpp,
    but it is only compatible with the upcoming Godot 4.0.
    Therefore, you need to use the ``master`` branch of godot-cpp to use GDExtension
    and follow this example.

If you are versioning your project using Git, it is a good idea to add them as
Git submodules:

.. code-block:: none

    mkdir gdextension_cpp_example
    cd gdextension_cpp_example
    git init
    git submodule add -b master https://github.com/godotengine/godot-cpp
    cd godot-cpp
    git submodule update --init

If you decide to just download the repositories or clone them into your project
folder, make sure to keep the folder layout identical to the one described here,
as much of the code we'll be showcasing here assumes the project follows this
layout.

> 

Reviewer 3: What folder structure?
Maybe insert a graphic.

Author: Not sure what you mean. I have code blocks further below that showcase the file structure.

> Do make sure you clone recursive to pull in both repositories:

Reviewer 2: ```suggestion
Do make sure you clone recursively to pull in both repositories:
```

To speed up compilation, add `-jN` at the end of the SCons command line where `N`
is the number of CPU threads you have on your system. The example below uses 4 threads.

.. code-block:: none

    cd godot-cpp
    scons platform=<platform> -j4
    cd ..

This step will take a while. When it is completed, you should have static
libraries that can be compiled into your project stored in ``godot-cpp/bin/``.

Creating a simple plugin
------------------------

Now it's time to build an actual plugin. We'll start by creating an empty Godot
project in which we'll place a few files.

Open Godot and create a new project. For this example, we will place it in a
folder called ``demo`` inside our GDExtension module's folder structure.

In our demo project, we'll create a scene containing a Node called "Main" and
we'll save it as ``main.tscn``. We'll come back to that later.

Back in the top-level GDExtension module folder, we're also going to create a
subfolder called ``src`` in which we'll place our source files.

You should now have ``demo``, ``godot-cpp``, and ``src``
directories in your GDExtension module.

Your folder structure should now look like this:

> 

Reviewer 3: I am not able to see the tree
![image](https://user-images.githubusercontent.com/20573784/218527554-8656d5db-5d7c-429c-a51e-75ddd5d6c03d.png)

Reviewer 3: PS: this is what I meant with "Maybe insert a graphic."

Author: I think I fixed it. There was a Line break missing

Reviewer 3: Yes its fixed

In the ``src`` folder, we'll start with creating our header file for the
GDExtension node we'll be creating. We will name it ``gdexample.h``:

.. code-block:: C++

    #ifndef GDEXAMPLE_H
    #define GDEXAMPLE_H

    #include <godot_cpp/classes/sprite2d.hpp>

    namespace godot {

    class GDExample : public Sprite2D {
        GDCLASS(GDExample, Sprite2D)

    private:
        float time_passed;

    protected:
        static void _bind_methods();

    public:
        GDExample();
        ~GDExample();

        void _init(); // our initializer called by Godot

        void _process(float delta);
    };

    }

    #endif

There are a few things of note to the above. We include ``sprite2d.hpp`` which
contains bindings to the Sprite2D class. We'll be extending this class in our
module.

We're using the namespace ``godot``, since everything in GDExtension is defined
within this namespace.

Then we have our class definition, which inherits from our Sprite2D through a
container class. We'll see a few side effects of this later on. The
``GDCLASS`` macro sets up a few internal things for us.

After that, we declare a single member variable called ``time_passed``.

> In the next block we're defining our methods, we obviously have our constructor

Reviewer 3: obviously?
A lot of people reading this will not be familiar with c++

Author: Good catch! Thx

        void _process(float delta);
    };

    }

    #endif

There are a few things of note to the above. We include ``sprite2d.hpp`` which
contains bindings to the Sprite2D class. We'll be extending this class in our
module.

We're using the namespace ``godot``, since everything in GDExtension is defined
within this namespace.

Then we have our class definition, which inherits from our Sprite2D through a
container class. We'll see a few side effects of this later on. The
``GDCLASS`` macro sets up a few internal things for us.

After that, we declare a single member variable called ``time_passed``.

In the next block we're defining our methods, we obviously have our constructor
and destructor defined, but there are two other functions that will likely look
familiar to some, and one new method.

The first is ``_bind_methods``, which is a static function that Godot will
call to find out which methods can be called and which properties it exposes. 
The second is our ``_process`` function, which will work exactly the same 
as the ``_process`` function you're used to in GDScript. The third is 
our ``_init`` function which is called after Godot has properly set up
our object.

> 

Reviewer 3: What is the difference beween extensions  ``_init``  and gdscript  ``_init``

Author: I am not sure to be quite honest. There was this dicussion comment here: https://github.com/godotengine/godot-cpp/issues/348#issuecomment-742084113 saying:
> _init is null because in the Godot API, _init is a """""virtual""""" function. Not exactly in the C++ sense though: this is a function that exists in the API because Godot expects us to be able to implement it optionally, but it doesn't actually define a default implementation itself. The function does not even exists in the engine (or sometimes it does under a different definition), it's purely a script API artifact. I have yet to find a class of the API with a virtual method you can actually call.

Reviewer 3: I would think all `Object`s are able to implement `_init` where the constructor is called when the cpp is created and the _init is called in the godot lifecycle. I would even know how to test the difference.
Let's keep it as is to remove confusion.

Reviewer 4: I'm not 100% sure anymore, it's been awhile, @Faless you may be able to confirm.

In GDNative `_init` had to be defined because the wrapper code always called it. I don't think this is so anymore in GDExternal, I think it's optional.
It should now work the same as in GDScript where you define `_init` and register it in `bind_methods` and it will get called if needed.

	    ClassDB::add_property("GDExample", PropertyInfo(Variant::FLOAT, "speed"), "set_speed", "get_speed");
    }

    void GDExample::GDExample() {
        // initialize any variables here
        time_passed = 0.0;
        amplitude = 10.0;
        speed = 1.0;
    }

    void GDExample::_process(float delta) {
        time_passed += speed * delta;

        Vector2 new_position = Vector2(
            amplitude + (amplitude * sin(time_passed * 2.0)),
            amplitude + (amplitude * cos(time_passed * 1.5))
        );

        set_position(new_position);
    }

    void GDExample::set_speed(float p_speed) {
        speed = p_speed;
    }

    float GDExample::get_speed() const {
        return speed;
    }

Now when the project is compiled, we'll see another property called speed.
Changing its value will make the animation go faster or slower.

.. note::

    For simplicity, we've left out the optional parameters in the
    add_property() method call. These parameters are
    ``hint``, ``hint_string``, ``usage`` and ``class_name``. These can be used to
    further configure how properties are displayed and set on the Godot side.

> 

Reviewer 3: add_property() could use a tutorial by itself if it is not already available

Author: There could be a more detailed tutorial about properties but this PR focuses on having a minimum working GDExtension first.

Reviewer 4: I don't know, seeing we do amplitude first, we could add additional values to speed to provide a range just to indicate one possible use case. We can then change this note to indicate there are more options.

That said, I'm all for having a separate page that purely handles properties and goes through the various things that are possible with these parameters as we really lack examples of each one.

Author: @Gallilus @BastiaanOlij 
Alright, I added a property hint for the `speed` property. This should do for a simple example. I might be able to go into further detail in a follow-up PR

Reviewer 4: Looks good, its enough to peek someones interest and give atleast one example of how to use hints.

submodules are not automatically initialized. You will need to execute the
following commands:

.. code-block:: none

    cd gdextension_cpp_example
    git submodule update --init --recursive

This will clone these two repositories into your project folder.

Building the C++ bindings
-------------------------

Now that we've downloaded our prerequisites, it is time to build the C++
bindings.

The repository contains a copy of the metadata for the current Godot release,
but if you need to build these bindings for a newer version of Godot, simply
call the Godot executable:

.. code-block:: none

    godot --dump-extension-api extension_api.json

Place the resulting ``extension_api.json`` file in the project folder and add
``custom_api_file=<PATH_TO_FILE>`` to the scons command
below.

To generate and compile the bindings, use this command (replacing ``<platform>``
with ``windows``, ``linux`` or ``macos`` depending on your OS):

To speed up compilation, add `-jN` at the end of the SCons command line where `N`
is the number of CPU threads you have on your system. The example below uses 4 threads.

.. code-block:: none

    cd godot-cpp

>     scons platform=<platform> -j4

Reviewer: ```suggestion
    scons platform=<platform> -j4 custom_api_file=<PATH_TO_FILE>
```

It makes it clearer for the reader where the `extension_api.json` is used

We first call a function in our bindings library that does its usual stuff.
After that, we call the function ``register_class`` for each of our classes in
our library.

At last, we need the header file for the ``register_types.cpp`` named
``register_types.h``.

.. code-block:: C++

    #ifndef GDEXAMPLE_REGISTER_TYPES_H
    #define GDEXAMPLE_REGISTER_TYPES_H

    void initialize_example_module();
    void uninitialize_example_module();

    #endif // GDEXAMPLE_REGISTER_TYPES_H

Compiling the plugin
--------------------

We cannot easily write by hand a ``SConstruct`` file that SCons would use for
building. For the purpose of this example, just use
:download:`this hardcoded SConstruct file <files/cpp_example/SConstruct>` we've
prepared. We'll cover a more customizable, detailed example on how to use these
build files in a subsequent tutorial.

.. note::

    This ``SConstruct`` file was written to be used with the latest ``godot-cpp``
    master, you may need to make small changes using it with older versions or
    refer to the ``SConstruct`` file in the Godot 4.0 documentation.

Once you've downloaded the ``SConstruct`` file, place it in your GDExtension module

> folder besides ``godot-cpp``, ``src`` and ``demo``, then run:

Reviewer: ```suggestion
folder beside ``godot-cpp``, ``src`` and ``demo``, then run:
```
Maybe you meant "alongside" or "beside". You can use "next to" as well

Introduction
------------

The C++ bindings for GDExtension are built on top of the C GDExtension API
and provide a nicer way to "extend" nodes in Godot using C++.
This new system allows the extension of Godot to nearly the same
level as statically linked C++ modules.

You can download the included example in the test folder of the godot-cpp
repository `on GitHub <https://github.com/godotengine/godot-cpp>`__.

Setting up the project
----------------------

There are a few prerequisites you'll need:

- a Godot 4 executable,
- a C++ compiler,
- SCons as a build tool,
- a copy of the `godot-cpp
  repository <https://github.com/godotengine/godot-cpp>`__.

See also :ref:`Compiling <toc-devel-compiling>` as the build tools are identical
to the ones you need to compile Godot from source.

You can download this repository from GitHub or let Git do the work for you.
Note that this repository has different branches for different versions
of Godot. GDExtensions will not work in older versions of Godot (only Godot 4 and up) and vice versa, so make sure you download the correct branch.

.. note::

    `GDExtension <https://godotengine.org/article/introducing-gd-extensions>`__

>     has been merged in the ``master`` branch of godot-cpp,

Reviewer 4: Hmm, this note is weird. It makes sense to have in the Godot 3 documentation to refer to needing to use a branch other than master in order to still do things with GDNative.

But I think using the master branch of godot-cpp is expected especially with a view of Godot 4 stable being released any day now.

I'd remove it or only keep `You need to use the ``master`` branch of godot-cpp to use GDExtension and follow this example.`

    but it is only compatible with the upcoming Godot 4.0.
    Therefore, you need to use the ``master`` branch of godot-cpp to use GDExtension
    and follow this example.

If you are versioning your project using Git, it is a good idea to add them as
Git submodules:

.. code-block:: none

    mkdir gdextension_cpp_example
    cd gdextension_cpp_example
    git init
    git submodule add -b master https://github.com/godotengine/godot-cpp
    cd godot-cpp
    git submodule update --init

If you decide to just download the repositories or clone them into your project
folder, make sure to keep the folder layout identical to the one described here,
as much of the code we'll be showcasing here assumes the project follows this
layout.

Do make sure you clone recursively to pull in both repositories:

.. code-block:: none

    mkdir gdextension_cpp_example
    cd gdextension_cpp_example
    git clone --recursive -b master https://github.com/godotengine/godot-cpp

.. note::

>     ``godot-cpp`` now includes ``godot-headers`` as a nested submodule, if you've

Reviewer 4: Need to double check this, I believe that in GDExtensions we no longer maintain the `godot-headers` repo and just maintain all files in `godot-cpp`.

I think you can thus also remove the mention of needing to recursively clone.

Might want to double check this but I think this is so.

Author: You are right about this. I will remove this then

``GDCLASS`` macro sets up a few internal things for us.

After that, we declare a single member variable called ``time_passed``.

In the next block we're defining our methods, we have our constructor
and destructor defined, but there are two other functions that will likely look
familiar to some, and one new method.

The first is ``_bind_methods``, which is a static function that Godot will
call to find out which methods can be called and which properties it exposes. 
The second is our ``_process`` function, which will work exactly the same 
as the ``_process`` function you're used to in GDScript. The third is 
our ``_init`` function which is called after Godot has properly set up
our object.

Let's implement our functions by creating our ``gdexample.cpp`` file:

.. code-block:: C++

    #include "gdexample.h"
    #include <godot_cpp/core/class_db.hpp>

    using namespace godot;

    void GDExample::_bind_methods() {
        ClassDB::bind_method(D_METHOD("_process"), &GDExample::_process);
    }

    GDExample::GDExample() {
        // initialize any variables here
        time_passed = 0.0;
    }

    GDExample::~GDExample() {
        // add your cleanup here
    }

>     void GDExample::_process(float delta) {

Reviewer 4: I think process works differently now. Process is now defined as a virtual in `node.hpp`: `virtual void _process(double delta);`.

So you now simply override it. You also don't need to bind it, it's already bound.

Reviewer 4: It's also important to call `_process` on the super class or any logic in the super classes will be skipped.

The same applies to a bunch of other functions, just have a look at `node.hpp`.

        time_passed += delta;

        Vector2 new_position = Vector2(10.0 + (10.0 * sin(time_passed * 2.0)), 10.0 + (10.0 * cos(time_passed * 1.5)));

        set_position(new_position);
    }

This one should be straightforward. We're implementing each method of our class
that we defined in our header file.

Note our ``_process`` function, which keeps track of how much time has passed
and calculates a new position for our sprite using a sine and cosine function.

> What stands out is calling ``owner->set_position`` to call one of the built-in

Reviewer 4: `owner` is no longer applicable I think? That was a GDNative thing.

Here is another overview to check the correct file structure:

.. code-block:: none

    gdextension_cpp_example/
    |
    +--demo/                  # game example/demo to test the extension
    |   |
    |   +--main.tscn
    |   |
    |   +--bin/
    |       |
    |       +--gdexample.gdextension
    |
    +--godot-cpp/             # C++ bindings
    |
    +--src/                   # source code of the extension we are building
    |   |
    |   +--register_types.cpp
    |   +--register_types.h
    |   +--gdexample.cpp
    |   +--gdexample.h

Time to jump back into Godot. We load up the main scene we created way back in
the beginning and now add a newly available GDExample node to the scene:

.. image:: img/gdextension_cpp_nodes.png

We're going to assign the Godot logo to this node as our texture, disable the
``centered`` property:

.. image:: img/gdextension_cpp_sprite.png

We're finally ready to run the project:

.. image:: img/gdextension_cpp_animated.gif

Adding properties
-----------------

GDScript allows you to add properties to your script using the ``export``
keyword. In GDExtension you have to register the properties with a getter and

> sett function or directly implement the ``_get_property_list``, ``_get`` and

Reviewer 4: `sett` -> `setter`

            amplitude + (amplitude * sin(time_passed * 2.0)),
            amplitude + (amplitude * cos(time_passed * 1.5))
        );

        set_position(new_position);
    }

    void GDExample::set_amplitude(const float p_amplitude) {
        amplitude = p_amplitude;
    }

    float GDExample::get_amplitude() const {
        return amplitude;
    }

Once you compile the module with these changes in place, you will see that a
property has been added to our interface. You can now change this property and
when you run your project, you will see that our Godot icon travels along a
larger figure.

Let's do the same but for the speed of our animation and use a setter and getter
function. Our ``gdexample.h`` header file again only needs a few more lines of
code:

.. code-block:: C++

    ...
        float amplitude;
        float speed;
    ...
        void _process(float delta) override;
        void set_speed(float p_speed);
        float get_speed();
    ...

This requires a few more changes to our ``gdexample.cpp`` file, again we're only
showing the methods that have changed so don't remove anything we're omitting:

.. code-block:: C++

    void GDExample::_bind_methods() {

>         ClassDB::bind_method(D_METHOD("get_speed"), &GDExample::get_speed);

Reviewer 4: might want to add some `...` to indicate we're not removing the code in the fragment above.

.. code-block:: C++

    void GDExample::_process(float delta) {
        time_passed += speed * delta;

        Vector2 new_position = Vector2(
            amplitude + (amplitude * sin(time_passed * 2.0)),
            amplitude + (amplitude * cos(time_passed * 1.5))
        );

        set_position(new_position);

        time_emit += delta;
        if (time_emit > 1.0) {
            emit_signal("position_changed", this, new_position);

            time_emit = 0.0;
        }
    }

After a second has passed, we emit our signal and reset our counter. We can add
our parameter values directly to ``emit_signal``.

Once the GDExtension library is compiled, we can go into Godot and select our sprite
node. In the **Node** dock, we can find our new signal and link it up by pressing
the **Connect** button or double-clicking the signal. We've added a script on
our main node and implemented our signal like this:

.. code-block:: GDScript

    extends Node

    func _on_Sprite2D_position_changed(node, new_pos):
        print("The position of " + node.get_class() + " is now " + str(new_pos))

Every second, we output our position to the console.

> 

Reviewer 4: Similar to what was discussed on properties, we really need a follow up page that goes more in depth on signals and specifically how to connect to signals from GDExtensions as the callables syntax is not something one would easily figure out on their own. But that should be a follow up PR
